src/server/services/procedures/cloud-variables/cloud-variables.js

```diff
@@ -396,21 +391,115 @@ CloudVariables.setUserVariable = function(name, value) {
+    delete (userListeners[username] || {})[name];
+};
+
+/**
+ * Equivalent to calling :func:`CloudVariables.lockVariable` followed by :func:`CloudVariables.getVariable`.
+ * @param {String} name Variable name
+ * @param {String=} password Password (if password-protected)
+ */
+CloudVariables.lockAndGetVariable = async function(name, password) {
+    await this.lockVariable(name, password);
+    return await this.getVariable(name, password);
+};
+
+/**
+ * Equivalent to calling :func:`CloudVariables.setVariable` followed by :func:`CloudVariables.unlockVariable`.
+ * @param {String} name Variable name
+ * @param {Any} value Value to store in variable
+ * @param {String=} password Password (if password-protected)
+ */
+CloudVariables.setAndUnlockVariable = async function(name, value, password) {
+    await this.setVariable(name, value, password);
+    await this.unlockVariable(name, password);
+};
+
+/**
+ * Sets the variable to the given value and returns the previous value.
+ * @param {String} name Variable name
+ * @param {Any} value Value to store in variable
+ * @param {String=} password Password (if password-protected)
+ */
+CloudVariables.getAndSetVariable = async function(name, value, password) {
+    const res = await this.getVariable(name, password);
+    await this.setVariable(name, value, password);
+    return res;
+};
+
+/**
+ * Sets the variable to the given value and returns the previous value.
+ * @param {String} name Variable name
+ * @param {Any} value Value to store in variable
+ */
+CloudVariables.getAndSetUserVariable = async function(name, value) {
+    const res = await this.getUserVariable(name);
+    await this.setUserVariable(name, value);
+    return res;
+};
+
+CloudVariables._getListenBucket = function (name) {
+    let bucket = globalListeners[name];
+    if (!bucket) bucket = globalListeners[name] = {};
+    return bucket;
+};
+CloudVariables._getUserListenBucket = function (name) {
+    const user = this.caller.username;
+    let userBucket = userListeners[user];
+    if (!userBucket) userBucket = userListeners[user] = {};
+
+    let bucket = userBucket[name];
+    if (!bucket) bucket = userBucket[name] = {};
+    return bucket;
+};
+
+/**
+ * Registers your client to receive messages each time the variable value is updated.
+ * ``name`` and ``password`` denote the variable to listen to.
+ * ``msgType`` is the name of the message that will be sent each time it is updated.
+ * 
+ * The variable must already exist prior to calling this RPC.
+ * Update events will cease when the variable is deleted.
+ * 
+ * **Message Fields**
+ * 
+ * - ``name`` - the name of the variable that was updated
+ * - ``value`` - the new value of the variable
+ * 
+ * @param {String} name Variable name
+ * @param {Any} msgType Message type to send each time the variable is updated
+ * @param {String=} password Password (if password-protected)
+ */
+CloudVariables.listenToVariable = async function(name, msgType, password) {
```

Comment:
Hmm... It might also be good to add a default/max duration for these so we can ensure we don't have an unbounded growth situation. (Probably ensure that a client can't be added twice, too.)

Reply:
The max duration sounds good - we could default to an hour or so, which should be fine for most all programs, and allow unbounded if they _really_ need it.

The adding client twice issue should be handled by having the final layer of map over client id.

Reply:
My concern with unbounded duration is also that server reboots would lose this info since it isn't persisted. If we want to allow unbounded, we would probably want to:
- store the socket handles in the db
- make sure we can clean up the listeners when the projects are closed